pkg/authorizer/token_retriever.go

```diff
@@ -12,63 +13,88 @@ import (
 	"time"
 
 	"github.com/Azure/msi-acrpull/pkg/authorizer/types"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/confidential"
 )
 
 const (
 	defaultARMResource              = "https://management.azure.com/"
 	customARMResourceEnvVar         = "ARM_RESOURCE"
 	msiMetadataEndpoint             = "http://169.254.169.254/metadata/identity/oauth2/token"
 	defaultCacheExpirationInSeconds = 600
+	authorityHost                   = "https://login.microsoftonline.com/"
+	resource                        = "https://management.azure.com/.default"
 )
 
-// TokenRetriever is an instance of ManagedIdentityTokenRetriever
-type TokenRetriever struct {
-	metadataEndpoint string
-	cache            sync.Map
-	cacheExpiration  time.Duration
+// TokenRetriever is an instance of ManagedIdentityTokenRetriever or WorkloadIdentityTokenTriever
+
+type BaseTokenRetriever struct {
+	cache           sync.Map
+	cacheExpiration time.Duration
+}
+type ManagedIdentityTokenRetriever struct {
+	metadataEndpoint   string
+	baseTokenRetriever *BaseTokenRetriever
+}
+
+type WorkloadIdentityTokenRetriever struct {
+	baseTokenRetriever *BaseTokenRetriever
 }
 
 type cachedToken struct {
 	token    types.AccessToken
 	notAfter time.Time
 }
 
+var baseTokenRetriever = NewBaseTokenRetriever()
+
+func NewBaseTokenRetriever() *BaseTokenRetriever {
+	return &BaseTokenRetriever{
+		cache:           sync.Map{},
+		cacheExpiration: time.Duration(defaultCacheExpirationInSeconds) * time.Second,
+	}
+}
+
 // NewTokenRetriever returns a new token retriever
-func NewTokenRetriever() *TokenRetriever {
-	return &TokenRetriever{
-		metadataEndpoint: msiMetadataEndpoint,
-		cache:            sync.Map{},
-		cacheExpiration:  time.Duration(defaultCacheExpirationInSeconds) * time.Second,
+func NewManagedIdentityTokenRetriever() *ManagedIdentityTokenRetriever {
+	return &ManagedIdentityTokenRetriever{
+		metadataEndpoint:   msiMetadataEndpoint,
+		baseTokenRetriever: baseTokenRetriever,
+	}
+}
+
+func NewWorkloadIdentityTokenRetriever() *WorkloadIdentityTokenRetriever {
+	return &WorkloadIdentityTokenRetriever{
+		baseTokenRetriever: baseTokenRetriever,
```

Comment:
See comment for `NewManagedIdentityTokenRetriever()`

Reply:
Done for next commit.